### PR TITLE
Fix race condition in access tests

### PR DESCRIPTION
**Pull request by sobolk**
## Problem

Access tests are using Amplify JS to obtain credentials which may lead to race conditions https://github.com/aws-amplify/amplify-backend/actions/runs/8842423407/job/24281918560#step:8:2674 in concurrent test settings.

**Issue number, if available:**

## Changes

This PR makes two changes:
1. Adds additional signOut before attempting to autheticate user.
2. Adds asynchronous lock to make sure that sequences of calls to AmplifyJs are atomic.

**Corresponding docs PR, if applicable:**

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/integration-tests/src/async_lock.ts
Comment (sobolk) [inline packages/integration-tests/src/async_lock.ts:18]: inspired by https://www.linkedin.com/pulse/asynchronous-locking-using-promises-javascript-abdul-ahad-o7smf/

File: packages/integration-tests/src/amplify_auth_credentials_factory.ts
Comment (awsluja) [inline packages/integration-tests/src/amplify_auth_credentials_factory.ts:73]: will this throw if a user is not logged in?

**Review by Amplifiyer — APPROVED** [review]

**Review by awsluja — APPROVED** [review]

File: packages/integration-tests/src/amplify_auth_credentials_factory.ts
Comment (sobolk) [reply packages/integration-tests/src/amplify_auth_credentials_factory.ts:73]: If checks pass on this PR then yes.
I'll wait for confirmation before merging. Or add some defensive logic otherwise.

**Pull request merged**
